R/cv_pensynth.R

```diff
@@ -57,38 +65,125 @@ cv_pensynth <- function(X1, X0, v, Z1, Z0, nlambda = 100, opt_pars = osqp::osqpS
 
   lseq <- lambda_sequence(X1VX0, Delta, nlambda)
 
-  # linear constraint matrix
-  Amat   <- rbind(rep(1, N_donors), diag(N_donors))
-  lbound <- c(1, rep(0, N_donors))
-  ubound <- rep(1, N_donors + 1)
+  # osqp and clarabel require constraints in differing formats.
+  # The following sets up the correct function call for each.
+  if (solver == "osqp") {
+    # Set default parameters for osqp
+    if (is.null(opt_pars)) {
+      opt_pars <- osqp::osqpSettings(polish = TRUE)
+    }
+
+    # linear constraint matrix
+    Amat <- rbind(rep(1, N_donors), diag(N_donors))
+    lbound <- c(1, rep(0, N_donors))
+    ubound <- rep(1, N_donors + 1)
+
+    # Define function for solving qp for a given lambda
+    solve_qp <- function(lambda) {
+      o <- capture.output({
+        # Instantiate the quadratic program solver
+        qpsolver <- osqp::osqp(
+          P = X0VX0,
+          q = -X1VX0 + lambda * Delta,
+          A = Amat,
+          l = lbound,
+          u = ubound,
+          pars = opt_pars
+        )
+
+        # solve
+        result <- qpsolver$Solve()
+
+        # Unload info from osqp result, so we can return it more easily later
+        for (info_col in names(result$info)) {
+          result[[info_col]] <- result$info[[info_col]]
+        }
 
-  w_path <- sapply(lseq, function(lambda) {
-    o <- capture.output({
-      # Instantiate the quadratic program solver
-      qpsolver <- osqp::osqp(
+        # Drop info now that we've copied it to the result
+        result$info <- NULL
+      })
+
+      return(result)
+    }
+  } else if (solver == "clarabel") {
+    # Set default parameters for clarabel
+    if (is.null(opt_pars)) {
+      opt_pars <- clarabel::clarabel_control()
+    }
+
+    # Constraint matrices
+    Amat <- rbind(
+      rep(1, N_donors), # Sum to 1 constraint
+      diag(N_donors), # Individ. weights lte 1 constraint
+      -diag(N_donors) # Individ. weights gte 0 constraint
```

Comment:
We would only need sum to 1 and greater than 0 right? Because that automatically implies individual weights are less than 1.